src/test/java/org/apache/commons/lang3/StringUtilsChompTest.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.lang3;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.infra.Blackhole;
+
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.commons.lang3.StringUtils.EMPTY;
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+import static org.apache.commons.lang3.StringUtils.length;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Thread)
+public class StringUtilsChompTest {
+    private static final String string1 = "\r";
+    private static final String string2 = "a";
+    private static final String string3 = "aa";
+    private static final String string4 = "aa\r";
+    private static final String string5 = "aa\n";
+    private static final String string6 = buildString6();
+    private static final String[] strings = buildStrings();
```

Comment:
The majority of these strings will not end with either CR (char 13) or LF (char 10). So you will not be chomping anything most of the time. I do not think that is what you want to test the performance of.

Reply:
> The majority of these strings will not end with either CR (char 13) or LF (char 10). So you will not be chomping anything most of the time. I do not think that is what you want to test the performance of.

No, it is exactly what I test for.
I use test 1-6 to show its performance with parameter who ends with \r or \n, and use this test s to show its performance with parameter who ends with other chars.

Reply:
OK. Then you should comment each test to identify the intention, or better yet just name the test:

```java
public String test1Old() {
public String test1New() {
public String test2Old() {
public String test2New() {
```
Becomes:
```java
public String test_CR_Old() {
public String test_CR_New() {
public String test_SingleChar_Old() {
public String test_SingleChar_New() {
```

From you latest results I would guess the 6% performance improvement for strings with nothing to chomp is probably due to not requiring a String.substring call to return a new instance. That has value.

Reply:
> OK. Then you should comment each test to identify the intention, or better yet just name the test:
> 
> ```java
> public String test1Old() {
> public String test1New() {
> public String test2Old() {
> public String test2New() {
> ```
> 
> Becomes:
> 
> ```java
> public String test_CR_Old() {
> public String test_CR_New() {
> public String test_SingleChar_Old() {
> public String test_SingleChar_New() {
> ```

good idea. I will try to rename the tests.

> From you latest results I would guess the 6% performance improvement for strings with nothing to chomp is probably due to not requiring a String.substring call to return a new instance. That has value.

Yes, I think most reasons be in String.substring. but String.substring do not create a new String instance when it using (0, length) in cases like this.
Still it have some checkings in that String.substring function, that might be what it cost the 6%